lib/derivation.ml

```diff
@@ -162,9 +162,23 @@ let rec derive context = function
         ])
     ) in
     results >>= fun(cs) -> Ok (beta, C.Conj [C.Disj cs; c_e])
-  | ListCons (_e1, _e2) ->
-     (* TODO: support cons expr *)
-     raise Known_error.(FialyzerError (NotImplemented {issue_links=["https://github.com/dwango/fialyzer/issues/90"]; message="support cons expr"}))
+  | ListCons (hd, tl) ->
+    let alpha = new_tyvar() in
+    let union_of elements1 elements2 = match (elements1, elements2) with
+    | (TyUnion elements1, TyUnion elements2) -> TyUnion (List.append elements1 elements2)
+    | (TyUnion elements, _) -> TyUnion elements
+    | (_, TyUnion elements) -> TyUnion elements
+    | (TyAny, TyAny) -> TyAny
+    | (TyAny, TyBottom) -> TyAny
+    | (TyBottom, TyBottom) -> TyBottom
+    | (TyBottom, TyAny) -> TyAny
+    in
+    derive context hd >>= fun (ty_hd, c_hd) ->
+    derive context tl >>= fun (ty_tl, c_tl) ->
+      Ok (Type.of_elem (TyList (union_of alpha ty_hd)), C.Conj [
```

Comment:
I think it would be better to use `sup` instead of `union_of`.

```suggestion
      Ok (Type.of_elem (TyList (Type.sup alpha ty_hd)), C.Conj [
```